src/backend/utils/sort/tuplestore.c

```diff
@@ -607,37 +644,40 @@ tuplestore_cleanup(Tuplestorestate *state, bool should_abort)
 
 		Assert(sstate->session_id == gp_session_id);
 		/*
-		 * If we are aborting, set the atomic variable in shared memory.
+		 * If we are aborting, set the variable in shared memory.
 		 * This must be done before everything else, since if we can't
 		 * delete the tuplestore right now, we expect other backends to
 		 * do so after they're done using it.
 		 */
 		if (should_abort)
-			pg_atomic_exchange_u32(&sstate->aborting, 1);
-		uint32 num_total_left = pg_atomic_sub_fetch_u32(&sstate->num_total_left, 1);
-		uint32 num_current = pg_atomic_sub_fetch_u32(&sstate->num_current, 1);
-		Assert(num_current <= num_total_left);
-		if (num_current == 0) /* safe to delete, no one is using it */
+			sstate->aborting = true;
+		sstate->num_done++;
+		sstate->num_current--;
+		Assert(sstate->num_done <= sstate->num_total);
+		if (sstate->num_current == 0) /* safe to delete, no one is using it */
 		{
 			/*
 			 * Either our backend is aborting, or some other backend was
 			 * and set the flag for us. It's this backend's job to delete
 			 * the files now.
 			 */
-			uint32 aborting = pg_atomic_read_u32(&sstate->aborting);
-			if (aborting || num_total_left == 0) /* should delete */
+			if (sstate->aborting || sstate->num_done >= sstate->num_total) /* should delete */
```

Comment:
just above there is an assert
```
Assert(sstate->num_done <= sstate->num_total);
```
so  situation `sstate->num_done > sstate->num_total` is not possible. Therefore

```suggestion
			if (sstate->aborting || sstate->num_done == sstate->num_total) /* should delete */
```

Reply:
Done